### PR TITLE
Validate the Resource Function Names at Compile Time

### PR DESCRIPTION
**Pull request by DimuthuMadushan**
## Purpose

Fixes: [#1270](https://github.com/ballerina-platform/ballerina-standard-library/issues/1270) 

## Examples
N/A

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java [hunk 1]
```diff
@@ -98,6 +102,19 @@ private void validateResourceAccessorName(MethodSymbol methodSymbol, Location lo
         }
     }
 
+    private void validateResourcePath(MethodSymbol methodSymbol, Location location, SyntaxNodeAnalysisContext context) {
+        if (methodSymbol.kind() == SymbolKind.RESOURCE_METHOD) {
+            ResourceMethodSymbol resourceMethodSymbol = (ResourceMethodSymbol) methodSymbol;
+            if (resourceMethodSymbol.resourcePath().signature().isEmpty()) {
+                PluginUtils.updateContext(context, CompilationErrors.INVALID_RESOURCE_PATH, location);
```
Comment (ThisaruGuruge) [inline graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java:109]: What does this condition mean?

Comment (pubudu91) [reply graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java:109]: I don't think we'll ever come across a scenario like this. A resource path will never be empty. If it is, that will be captured in the compiler itself.

File: graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java [hunk 2]
```diff
@@ -113,6 +130,19 @@ private void validateReturnType(TypeSymbol returnTypeDesc,
             if (typeReferenceTypeSymbol.definition().kind() == SymbolKind.TYPE_DEFINITION) {
                 TypeDefinitionSymbol typeDefinitionSymbol =
                         (TypeDefinitionSymbol) typeReferenceTypeSymbol.definition();
+                if (typeReferenceTypeSymbol.typeDescriptor().typeKind() == TypeDescKind.RECORD) {
+                    RecordTypeSymbol recordTypeSymbol = (RecordTypeSymbol) typeReferenceTypeSymbol.typeDescriptor();
+                    Map<String, RecordFieldSymbol>  recordFieldSymbolMap =  recordTypeSymbol.fieldDescriptors();
+                    for (Map.Entry<String, RecordFieldSymbol> entry:recordFieldSymbolMap.entrySet()) {
+                        if (entry.getValue().typeDescriptor().typeKind().equals(TypeDescKind.TYPE_REFERENCE)) {
```
Comment (pubudu91) [inline graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java:137]: Can use `==` here like done in the other places.

File: graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java [hunk 3]
```diff
@@ -113,6 +130,19 @@ private void validateReturnType(TypeSymbol returnTypeDesc,
...
+                            validateReturnType(entry.getValue().typeDescriptor(), location, context);
+                        } else {
+                            if (PluginUtils.isInvalidPath(entry.getKey())) {
```
Comment (pubudu91) [inline graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/GraphqlFunctionValidator.java:140]: This seems to be a generic check, not something specific to resource paths? If that's the case, shall we rename this method? It's a bit misleading

File: graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+type Lift record {
+    string __id;
+    Status getStatus;
+};
+
+service class Trail {
+
+    isolated resource function get id () returns string {
```
Comment (aashikam) [inline graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal:32]: Not sure if we keep spaces between the accessor name (`id`) and parentheses `()`

Comment (aashikam) [reply graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal:32]: Then again, it's fine for a test I guess

Comment (ThisaruGuruge) [reply graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal:32]: Let's remove the space here.

File: graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/PluginConstants.java [hunk 5]
```diff
@@ -61,7 +62,9 @@
                 "http:Listener and graphql:ListenerConfiguration are mutually exclusive", "GRAPHQL_110",
                 DiagnosticSeverity.ERROR),
         INVALID_RETURN_TYPE_MULTIPLE_SERVICES("Invalid return type with multiple indistinct services",
-                "GRAPHQL_111", DiagnosticSeverity.ERROR);
+                "GRAPHQL_111", DiagnosticSeverity.ERROR),
+        INVALID_RESOURCE_PATH("Invalid field name. GraphQL field Name must not begin with \"__\", " +
```
Comment (ThisaruGuruge) [inline graphql-compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/PluginConstants.java:66]: This name is misleading. Shall we rename this to something like `INVALID_FIELD_NAME`? 
```suggestion
        INVALID_FIELD_NAME("Invalid field name. A GraphQL field Name must not begin with \"__\", " +
```

**Review by ThisaruGuruge — APPROVED** [review]
LGTM

File: graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+service class Trail {
+
+    isolated resource function get id() returns string {
+        return "T1";
+    }
+
+    isolated resource function get __name() returns string {
+            return "Trail1";
+        }
+
```
Comment (ThisaruGuruge) [inline graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal:39]:
```suggestion
    isolated resource function get __name() returns string {
        return "Trail1";
    }
```

File: changelog.md [hunk 7]
```diff
@@ -37,6 +37,7 @@ This file contains all the notable changes done to the Ballerina GraphQL package
 - [[#1368] Fix Crashing the Schema Generation when a Union Type not Having a Name](https://github.com/ballerina-platform/ballerina-standard-library/issues/1368)
 - [[#1370] Fix Service Hanging When Returning an Enum Array](https://github.com/ballerina-platform/ballerina-standard-library/issues/1370)
 - [[#1391] Fix Union of Service Types Returning all the Members](https://github.com/ballerina-platform/ballerina-standard-library/issues/1391)
+- [[#1270] Add Compile Time Validation for Resource Function Names](https://github.com/ballerina-platform/ballerina-standard-library/issues/1270)
```
Comment (ThisaruGuruge) [inline changelog.md:40]: Shall we move this to the `Added` section?

File: graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    isolated resource function get lift() returns Lift {
+        return {
+                   __id:"L1",
+                   getStatus: {
+                       capacity:10,
+                       __elevationgain:20,
+                       night:false
+                   }
+               };
```
Comment (ThisaruGuruge) [inline graphql-compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_17/service.bal:58]: Shall we indent these properly?

**Comment by ThisaruGuruge** [comment]
Tests are failing. Can you sync with the master and check for the test failures too?

**Pull request merged**